### PR TITLE
Added LOG_PREAMBLE for custom log preamble strings

### PR DESCRIPTION
**Pull request by ErezBinyamin**
This should address Issue #8 which is the feature request for "Cutsom log format - add timestamp"

# What I changed
```C
// Before the log macros looked like this
#define LOG_ERROR(...) DebugLog::Manager::get().log(arx::debug::LogLevel::LVL_ERROR, LOG_SHORT_FILENAME, __LINE__, __func__, __VA_ARGS__)

// Now it looks like this
#define LOG_ERROR(...) DebugLog::Manager::get().log(arx::debug::LogLevel::LVL_ERROR, LOG_SHORT_FILENAME, LOG_PREAMBLE, __VA_ARGS__)
```

# How to use it
Before you `#include <DebugLog.h>` simply define a LOG_PREAMBLE macro.
```C
#define LOG_PREAMBLE __LINE__, __func__, "This string will go before all log statements: "
#include <DebugLog.h>
```

### CODE REVIEW

File: DebugLog.h [hunk 1]
```diff
@@ -21,6 +21,12 @@ using DebugLogLevel = arx::debug::LogLevel;
 using DebugLogBase = arx::debug::LogBase;
 using DebugLogPrecision = arx::debug::LogPrecision;
 
+// Defualt log level to INFO
+#if !defined(DEBUGLOG_DEFAULT_LOG_LEVEL_ERROR) && !defined(DEBUGLOG_DEFAULT_LOG_LEVEL_WARN) && !defined(DEBUGLOG_DEFAULT_LOG_LEVEL_INFO) && !defined(DEBUGLOG_DEFAULT_LOG_LEVEL_DEBUG) && !defined(DEBUGLOG_DEFAULT_LOG_LEVEL_TRACE)
+  #warning "Defaulting log level to: DEBUGLOG_DEFAULT_LOG_LEVEL_INFO"
+  #define DEBUGLOG_DEFAULT_LOG_LEVEL_INFO
+#endif
+
```
Comment (hideakitai) [inline DebugLog.h]: We have `LOG_SET_LEVEL(level)` now, so this is not necessary for this pr (this p-r is for `LOG_PREAMBLE`, I think)

File: DebugLog/Manager.h [hunk 2]
```diff
@@ -401,27 +401,6 @@ namespace debug {
                 case LogLevel::LVL_TRACE: header = "[TRACE] "; break;
                 default: header = ""; break;
             }
-#ifdef ARDUINO
-            if (b_file) header += file + string_t(" ");
-            if (b_line) header += string_t("L.") + line + string_t(" ");
-            if (b_func) header += func + string_t(" ");
```
Comment (hideakitai) [inline DebugLog/Manager.h]: `b_file`, `b_line`, `b_func` are not used now? If so, it is good to remove them.

File: DebugLogEnable.h [hunk 3]
```diff
@@ -6,12 +6,32 @@
...
+#elif defined(DEBUGLOG_DEFAULT_LOG_LEVEL_WARN)
+  #define LOG_INFO(...)
+  #define LOG_DEBUG(...)
+  #define LOG_TRACE(...)
+#elif defined(DEBUGLOG_DEFAULT_LOG_LEVEL_INFO)
+  #define LOG_DEBUG(...)
+  #define LOG_TRACE(...)
+#elif defined(DEBUGLOG_DEFAULT_LOG_LEVEL_DEBUG)
+  #define LOG_TRACE(...)
+#endif
```
Comment (hideakitai) [inline DebugLogEnable.h]: This is not necessary for this pr, I think (this p-r is for `LOG_PREAMBLE`)

**Review by hideakitai — CHANGES_REQUESTED** [review]
Thanks for your contribution! Please take a look at my comments.

File: DebugLog/Manager.h (hunk 2 above)
Comment (ErezBinyamin) [reply DebugLog/Manager.h]: I agree, should be good to remove

File: DebugLogEnable.h (hunk 3 above)
Comment (ErezBinyamin) [reply DebugLogEnable.h]: Agreed, not necessary for this PR. This change (along with the earlier L24-28 DebugLog.h) should be in a different PR as they are unrelated to LOG_PREAMBLE

**Comment by ErezBinyamin** [comment]
> Thanks for your contribution! Please take a look at my comments.

To keep this PR on topic to LOG_PREAMBLE, I think we should use commit: [f23e880](https://github.com/hideakitai/DebugLog/pull/9/commits/f23e880328a25882a760ffd0a71577f90b2f55b0) and ignore changes in [9dcd178](https://github.com/hideakitai/DebugLog/pull/9/commits/9dcd178327d19458112af353c48eb5df713aa338) and [c4d67bc](https://github.com/hideakitai/DebugLog/pull/9/commits/c4d67bc619ade52aa56c05c315eb26be84e84bc4), then resolve conflicts in DebugLog/Manager.h

**Comment by ErezBinyamin** [comment]
Requested changes have been made. Things seem to be in line a bit more now, let me know what you think.

File: DebugLog/Manager.h (hunk 2 above)
Comment (hideakitai) [reply DebugLog/Manager.h]: Could you remove `b_file`, `b_line`, `b_func` and `option()` as well?

**Comment by hideakitai** [comment]
It would be great if you could update README about `LOG_PREAMBLE` and related changes :)

File: DebugLog/Manager.h (hunk 2 above)
Comment (ErezBinyamin) [reply DebugLog/Manager.h]: Made changes and tested on my device. Works on my end.

**Comment by ErezBinyamin** [comment]
> It would be great if you could update README about `LOG_PREAMBLE` and related changes :)

Done

File: DebugLogEnable.h [hunk 4]
```diff
@@ -6,12 +6,16 @@
 #undef ASSERT
 #undef ASSERTM
 
+#ifndef LOG_PREAMBLE
+  #define LOG_PREAMBLE LOG_SHORT_FILENAME, __LINE__, __func__
+#endif
```
Comment (hideakitai) [inline DebugLogEnable.h]: [nit] After all this time, I prefer the default output to remain the same as before :) (I may change it myself later, please understand!)

File: README.md [hunk 5]
```diff
@@ -201,6 +206,9 @@ You can disable `LOG_XXXX` and `ASSERT` macro completely by defining following m
 // You can also set default log level by defining macro (default: INFO)
 // #define DEBUGLOG_DEFAULT_LOG_LEVEL_TRACE
 
+// You can call other macros, functions, or define strings that will be included in all log statements
+#define LOG_PREAMBLE __FILE__ " : "
+
```
Comment (hideakitai) [inline README.md]: [nit] I prefer to write examples of PREAMBLE configurations with various patterns :) (please note that I may modify this myself later!)

**Review by hideakitai — APPROVED** [review]
I wrote some minor comments, but LGTM :+1: Thanks for your contribution! Please merge yourself in your own time.

File: DebugLogEnable.h (hunk 4 above)
Comment (ErezBinyamin) [reply DebugLogEnable.h]: No problem, mind specifying exactly what you'd like it to look like?

Is this the what you mean with "Same as before"?
`[LEVEL] FILE L.# FUNC : message...`

Which would mean something like:
`#define LOG_PREAMBLE LOG_SHORT_FILENAME, "L.", __LINE__, " ", __func__, ":"`

**Comment by ErezBinyamin** [comment]
> I wrote some minor comments, but LGTM +1 Thanks for your contribution! Please merge yourself in your own time.

Lets make sure everything is exactly as you like it before we merge. This is your project and I do not like leaving tasks behind for others to pick up. 

TODO:

- [x] `DebugLogEnable.h`: Default output should be "same as before"
- [x] `README.md`: examples of PREAMBLE configurations with various patterns

File: README.md (hunk 5 above)
Comment (ErezBinyamin) [reply README.md]: How about I provide examples of combinations of:

- [x] Strings
- [x] Macros
- [x] Function Calls

Would these three examples be enough?

File: DebugLogEnable.h (hunk 4 above)
Comment (ErezBinyamin) [reply DebugLogEnable.h]: Done

**Comment by ErezBinyamin** [comment]
Default Logging style has been restored. `README.md` updated with better examples and an explanation of the default option for `LOG_PREAMBLE`

**Comment by hideakitai** [comment]
Great! Thank you for your contribution!

**Pull request merged**